core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -339,21 +350,30 @@ protected long getBlockSizeAllocation(long pos) {
    * @return the {@link InputStream} for the UFS
    * @throws IOException if the stream cannot be created
    */
-  protected InputStream createUnderStoreBlockInStream(long blockStart, long length, String path)
-      throws IOException {
-    return new UnderStoreBlockInStream(mContext, blockStart, length, mBlockSize,
-        getUnderStoreStreamFactory(path, mContext));
-  }
-
-  protected UnderStoreStreamFactory getUnderStoreStreamFactory(String path, FileSystemContext
-      context) throws IOException {
+  protected InputStream createUnderStoreBlockInStream(long blockId, long blockStart, long length,
+      String path) throws IOException {
     if (Configuration.getBoolean(PropertyKey.USER_UFS_DELEGATION_ENABLED)) {
-      return new DelegatedUnderStoreStreamFactory(context, path);
+      try {
+        WorkerNetAddress address = mUfsReadLocationPolicy.getWorkerForBlock(
+            BlockLocationPolicyGetWorkerOptions.defaults()
+                .setBlockWorkerInfos(mBlockStore.getWorkerInfoList()).setBlockId(blockId)
+                .setBlockSize(length));
+        return StreamFactory
+            .createUfsBlockInStream(mContext, path, blockId, length, blockStart, address,
+                mInStreamOptions);
+      } catch (AlluxioException e) {
+        throw new IOException(e);
+      }
     } else {
-      return new DirectUnderStoreStreamFactory(path);
+      return new UnderStoreBlockInStream(mContext, blockStart, length, mBlockSize,
+          getUnderStoreStreamFactory(path));
     }
   }
 
+  protected UnderStoreStreamFactory getUnderStoreStreamFactory(String path) throws IOException {
```

Comment:
shouldn't this be `getDirectUnderStoreStreamFactory`? or maybe we should even inline this method